src/prof_recent.c

```diff
@@ -506,33 +580,65 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count = prof_recent_alloc_dump_prepare_locked(tsd);
+	ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	prof_recent_t *dump_last = NULL;
 
-	emitter_json_array_kv_begin(&emitter, "recent_alloc");
-	for (prof_recent_t *n = prof_recent_alloc_begin(tsd);
-	    n != prof_recent_alloc_end(tsd);
-	    n = prof_recent_alloc_next(tsd, n)) {
-		prof_recent_alloc_dump_node(&emitter, n);
+	buf_writer_set_allow_auto_flush(&buf_writer, false);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy_reserve;
+	for (prof_recent_t *node = dump_dummy->next; node != dump_dummy;
+	    node = node->next) {
+		buf_writer_commit(&buf_writer);
```

Comment:
Why commit at the beginning of the loop? It's a bit counter intuitive.

Reply:
Because we always want to have the closest point to revert to if needed. What would be an alternative?

Reply:
Maybe this way? It feels we should always have a "clean" state when starting a new iteration.

    if (buf_writer_overflowed(...)) {
        ...
    } else {
        commit();
    }

Reply:
I think the main difference is the handling for the first node: if it overflows, then where should we revert to? That's dependent on the implementation of the buffered writer. In my current implementation, the caller would have to explicitly call `commit()` to set the commit point.

Reply:
I see. Let's keep as is.

Reply:
Yeah, and also, if we put the `commit()` at the end, then the last `commit()` would not be useful. So `commit()` might not be a very good name: it may sound like we're committing what we have just done, but actually it's more for having a place to go back to before we do the next thing. Thoughts on a better name?

Reply:
Agreed. Maybe `_set_checkpoint`? That way calling at the beginning is intuitive.